lambdaclass/mirra_backend
# [GH-321] Silence Status Effect

Closes https://github.com/lambdaclass/champions_of_mirra/issues/321

## Motivation

Use Tags to implement silence status effects. Silenced units cannot cast their ultimate skills.

Fixes an issue with modifier and tag durations:
- For durations, we were not substracting the initial steps remaining by 1, which made them last 1 step longer than they should. An explenation for the logic behind why this happens has been added in the code.
- For tags, we were actually doing this, but the implementation was faulty: when we had a permanent duration, we substracted 1 from the -1, resulting in an unhandled -2 steps_remaining value.

## Summary of changes

- Implemented Silence.
- Added tests for Silence. 
- Fixed some issues with durations that were made evident by the newly added tests.
- Fixed Stun (was missing the "ControlEffect" prefix in the name

## How to test it?

Give a unit a skill that gives all enemies and all allies a permanent "StatusEffect.Silence" tag (effects can be made permanent by not specifying a "duration" key in its type). No ultimate skill usages should ever be logged in battle (you can ctrl+f "casting Ultimate skill" in your console). Remember to change `Logger.info` calls to `IO.inspect` in the code if your logs are getting cut out.
